File: handy-java/TypeExample.java
import java.util.List;
import java.util.Map;

/**
 * TYPE EXAMPLES
 *
 * Pro tip:
 * TypeScript union types often map to Java `enum` when the values are finite.
 * Mentioning this mapping in interviews shows language-transfer strength.
 */
public class TypeExample {
    enum UserRole {
        STUDENT,
        PARENT
    }

    record User(String name, int id) {}

    static String formatSummary(String name, int lessons, boolean beginner) {
        return name + " has " + lessons + " lessons. Beginner friendly: " + beginner;
    }

    public static void main(String[] args) {
        String courseName = "Java Basics";
        int lessonCount = 4;
        boolean beginnerFriendly = true;

        UserRole role = UserRole.PARENT;

        System.out.println(formatSummary(courseName, lessonCount, beginnerFriendly));
        System.out.println("Current role: " + role);

        User user = new User("Hayes", 0);
        Map<String, Object> userAsMap = Map.of(
                "name", user.name(),
                "id", user.id()
        );

        for (Map.Entry<String, Object> entry : userAsMap.entrySet()) {
            System.out.println(entry.getKey() + " -> " + entry.getValue());
        }

        for (String key : userAsMap.keySet()) {
            System.out.println(key + ": " + userAsMap.get(key));
        }

        List<User> users = List.of(
                new User("Alice", 1),
                new User("Bob", 2),
                new User("Charlie", 3)
        );
        System.out.println("User count: " + users.size());
    }
}

File: handy-ts/quiz/type.example.ts
/**
 * TYPE EXAMPLES
 *
 * Key ideas:
 * 1) Primitive types: string, number, boolean
 * 2) Union type: value can be one of multiple types
 * 3) Type alias: reusable custom type
 */

const courseName: string = "TypeScript Basics";
const lessonCount: number = 4;
const isBeginnerFriendly: boolean = true;

type UserRole = "student" | "parent";
const role: UserRole = "parent";

function formartSummary(name: string, lessons: number, beginner: boolean): string {
  return `${name} has ${lessons} lessons. Beginner friendly: ${beginner}`;
}

console.log(formartSummary(courseName, lessonCount, isBeginnerFriendly));
console.log(`Current role: ${role}`); 

// use Object.entries to loop object
interface User {
  name: string;
  id: number;
}

const user: User = {
  name: "Hayes",
  id: 0,
};

for (const[key, value] of Object.entries(user)){
  console.log(`${key}->${value}`)
}
//The for...in loop iterates over all enumerable properties of an object
for (const key in user) {
  if (Object.prototype.hasOwnProperty.call(user, key)) {
    // Type assertion or type guard is needed for type safety
    console.log(`${key}: ${user[key as keyof User]}`);
  }
}


const users: User[] = [
  { name: 'Alice', id: 1 },
  { name: 'Bob', id: 2 },
  { name: 'Charlie', id: 3 }  
];